Feature/crawler arch

iggi42 opened the pull request:
got all tests positive on my local machine.

Commit:
setup task_manager to ratelimit crawls

Commit:
setup task_manager to ratelimit crawls

Commit:
TODO kill a search task not needed anymore

Commit:
TODO kill a search task not needed anymore

Commit:
which children will be used is not defined yet

Commit:
which children will be used is not defined yet

Commit:
the interal state of TaskManager is now a struct (rather than a tuple)

Commit:
the interal state of TaskManager is now a struct (rather than a tuple)

Commit:
allow multiple task managers running at the same time

Commit:
allow multiple task managers running at the same time

Commit:
add docs and typespec to task_manager

Commit:
add docs and typespec to task_manager

Commit:
fix typespec

Commit:
fix me being stupid

Commit:
implement first basic task_manager test

Commit:
make task_manager working again, implement very basic test for search

Commit:
implement the test to don't hang on unending tasks

it is strange that TaskManager already passes this test.

Commit:
Merge branch 'feature/crawler_arch' of github.com:Belethors-General-Mods/belethor into feature/crawler_arch

Commit:
fix typo

Commit:
make stuff easier to understand (just a little)

Commit:
implement test "allow multiple task managers ..."

Commit:
speed up test

Commit:
add assert

Commit:
improve docs and logging a little bit

Commit:
rename internal state

Commit:
let it crash

Commit:
improve task_manager tests

Commit:
fixed unending task problem and reduce code redundancy

Commit:
make ebert happy #1

Commit:
fix unending task test for task_manager

Commit:
add small module_doc to task_manager.ex

Commit:
mix format

Commit:
mix format

Commit:
Merge branch 'feature/crawler_arch' of github.com:Belethors-General-Mods/belethor into feature/crawler_arch

Commit:
uncommit .gitignore

Commit:
make debug logging lazy

skwerlman commented on apps/crawler/test/task_manager_test.exs:
```diff
@@ -0,0 +1,101 @@
+    end
+
+    queue = [block, block, block, valid]
+    expected = [:timeout, :timeout, :timeout, input]
```
this should be merged with line 65 for consistency with the other tests

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+
+      {{:value, {pid, args}}, q} ->
+        if max_reached?(state) do
+          start_task(state.supervisor, pid, args)
```
see comment on line 61

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+    # otherwise queue it
+    if max_reached?(state) do
+      Logger.debug(fn -> "directly start task #{inspect(args)}" end)
+      start_task(state.supervisor, client, args)
```
Since the task we start will always be discarded locally (it only matters to the client, not us) we should probably discard the `%Task{}` in `start_link` and return a constant like `:ok` instead. The reason for that is it lets the `%Task{}` be discarded sooner, and it stops dialyxir from whining about `:unmatched_return`.

Also, these calls should be prefixed with `:ok = ` for dialyxir.

skwerlman commented on apps/crawler/lib/crawler/application.ex:
```diff
@@ -7,10 +7,7 @@ defmodule Crawler.Application do
-      # Starts a worker by calling: Crawler.Worker.start_link(arg)
-      # {Crawler.Worker, arg},
-    ]
+    children = []
```
Where is the TaskManager started if not here?

iggi42 replied:
not at all, right now.
I am uncertain wether we want one task manager for each provider (steam, nexus, beths, etc) or one for all.

skwerlman replied:
i think the simplest way is to have a task manager for each provider where rate limiting is a concern. that way we can have a simple 1-1 correspondence between task manager and rate limiter, and not have to worry about one site's rate limit blocking other sites

other providers that don't care about rate limiting can share a task manager with multiple workers

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+  end
+
+  @doc """
+  execute `provider.search(query)` in an rate limited way.
```
nitpick: 'an' -> 'a'

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+  If the task maximum is reached, the request will be queued.
+  """
+
+  defstruct [:max, :queue, :supervisor]
```
without really looking for it, its not obvious why there's a struct here, so this probably deserves a `typespec t` and a `typedoc` to explain it

iggi42 replied:
it is the internal state of gen_server, if at all a `@typedoc false`

skwerlman replied:
ye, but we still want good documentation, even for internal stuff. i'm fine with `@doc false` for callbacks, and leaving trivial `defp`s undocumented, but anything complex or non-obvious really needs documentation for maintainability reasons

iggi42 replied:
but the struct does (and should) only matter if task_manager.ex it self is changed.
a code comment should do it thus, right?

skwerlman replied:
a code comment could work, ye. but personally i find being able to see the internal documentation as a nice easy-to-navigate webpage (as `file:///mnt/code/belethor/doc/api-reference.html`) is preferable to reading comments. ofc this_is_ a personal preference, so if you think its not worth the effort i'm fine with dropping the issue

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+      supervisor: supervisor
+    }
+
+    Logger.debug(fn ->
```
this is an annoying side-effect of having dialyxir look for unmatched returns, but Logger calls should be prefixed with `_ =` so dialyxir knows we mean to ignore them

iggi42 replied:
I did an debug macro to handle all the special care Logger.debug needs.

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+    # startup a new task if the max is not reached
+    # otherwise queue it
+    if max_reached?(state) do
+      Logger.debug(fn -> "directly start task #{inspect(args)}" end)
```
prefix with `_ =`

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+      start_task(state.supervisor, client, args)
+      {:noreply, state}
+    else
+      Logger.debug(fn -> "request #{inspect(args)} will be queued" end)
```
prefix with `_ =`

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+  # a task returned without error
+  @doc false
+  def handle_info({ref, :ok}, state) when is_reference(ref) do
+    Logger.debug(fn -> "a Task (#{inspect(ref)}) ended successful" end)
```
prefix with `_ =`

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+  # a monitored process died (for whatever reason)
+  @doc false
+  def handle_info(down = {:DOWN, _ref, :process, _pid, _reason}, state) do
+    Logger.debug(fn -> "got a down msg : #{inspect(down)}" end)
```
prefix with `_ =`

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+
+  defp start_task(supervisor, client = {client_pid, _id}, {provider, query})
+       when is_pid(client_pid) do
+    Logger.debug(fn -> "current supervised children #{inspect(count_tasks(supervisor))}" end)
```
prefix with `_ =`

skwerlman commented on apps/crawler/test/task_manager_test.exs:
```diff
@@ -0,0 +1,101 @@
+    end
+  end
+
+  test "basic test ensure provider.search(query) gets called" do
```
nitpick: rename to `ensure provider.search(query) gets called`

skwerlman commented on apps/crawler/test/task_manager_test.exs:
```diff
@@ -0,0 +1,101 @@
+    input = ["basic testvalue"]
+    {:ok, mng} = TaskManager.start_link(1)
+    result = TaskManager.search(input, mng, EchoProvider)
+    assert result == input
```
this test should use the same `expected/actual` variables as the other tests

skwerlman commented on apps/crawler/test/task_manager_test.exs:
```diff
@@ -0,0 +1,101 @@
+        TaskManager.search(input, mng, EchoProvider, timeout)
+      catch
+        :exit, msg = {reason, _} ->
+          Logger.warn("valid request failed with #{inspect(msg)}")
```
it's probably better to raise here rather than simply log, since this should never happen

iggi42 replied:
well, it should happen if TaskManager.search exits ungracefully.
in that case reason is never `:ok` and the assert fails.
this code is just to keep the failing assertion clearly and the full exit details in the log.

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,118 @@
+
+  # callbacks and internal stuff
+  @doc false
+  def init(max) do
```
dialyxir complains here of the function having no return. I _think_ its because `init` is underspecced since dialyxir can't infer the success typing for `max`. Not 100% certain of that though.

if its the only remaining issue, then i'm ok with adding it to an ignore file and accepting the pr as long as we circle back later to fix it.

iggi42 replied:
sorry, I can't reproduce this. How do I make `mix dialyzer` show warnings?

skwerlman commented on apps/crawler/test/task_manager_test.exs:
```diff
@@ -0,0 +1,101 @@
+
+    assert expected == actual
+
+    Process.sleep(400)
```
why are these sleeps here? is it to work around some exunit weirdness, or..?

iggi42 replied:
to prevent the still queued log messages to be associated with the next test.
`Logger.flush` is a better choice

skwerlman commented:
~~the build failure is fine; probably a travis issue, not ours.~~ rerunning the build fixed it. we might need to fix the build script to avoid this, or we can just rerun every time

Commit:
fix skwerlman reclamtions

well, almost all

iggi42 commented:
I fixed most of the points you pointed out.
Also I am in the process of defining a behaviour for search providers.

Commit:
add the SearchProvider behaviour

Commit:
externalize task_supervisor from task_manager (API change)

this should allow prober allow supervision trees. (I guess)

Commit:
add a little documentation to search_provider

iggi42 commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,122 @@
+  ## api
+
+  @doc "start an instance, with options"
+  @spec start_link(max(), Supervisor.on_start(), GenServer.options()) :: Supervisor.on_start()
```
well `mix format` and ebert disagree on each other.

Commit:
fix typo

Commit:
make credo and format friends again, they ignore issues in the database for now

Commit:
another api change in task_manager.ex

Commit:
rename search provider to client

provider =~ nexus, steam, bethesda
client =~ code to access a api

so each provider has a client, but a provider is also more than just that client.
that is why I renamed the current search behaviour stuff to client

Commit:
start a task manager and its supervisor at crawler application start for each provider

Commit:
fix dialyzer and credo errors

Commit:
reword a test describtion

Commit:
fix shit

Commit:
register supervisors with a name for real

Commit:
add boilerplate code to use providers

Commit:
add alias and some (bad) @docs

Commit:
fix credos complain about a pipe

Commit:
improve docs and fix some small credo stuff

iggi42 commented:
The clients to nexus, steam and bethesda are not implemented yet.
Some type are not done correct yet, but I will wait with that until we have the common app to house some types.
I am not sure if I missed to implement some API to allow inter app communication.

If we fix these issues too, crawler should be done.

Commit:
credo thinks 2 alias should be alphabetical sorted

Commit:
add test to ensure task manager respects max tasks

Commit:
added test for the provider instaces

iggi42 committed:
Merge branch 'master' into feature/crawler_arch

Commit:
reformat with new max line length

Commit:
how did I miss that?

Commit:
remove multi-alias thingy and remove a line

skwerlman commented on apps/crawler/lib/crawler/client_behaviour.ex:
```diff
@@ -0,0 +1,11 @@
+defmodule Crawler.Client do
+  @moduledoc """
+  beaviour to implement search providers for `Crawler.TaskManager`
```
beaviour -> behaviour (or behavior)

actually, (and this is a question for another pr) we should decide on british vs american spellings at some point. for now it doesn't matter though

skwerlman replied:
also, IMO implement -> describe

i say this because a behavior lays out how an implementation should be written rather than actually implementing anything itself

skwerlman commented on apps/crawler/lib/crawler/provider_behaviour.ex:
```diff
@@ -0,0 +1,37 @@
+  @callback search!(Client.query()) :: [Client.result()]
+
+  defmacro __using__(args) do
+    name = args[:name]
```
pattern match in the function definition rather than parsing args. that way you get a compile-time error if you forget the name, not a run-time error

iggi42 replied:
args in `__using__` is a keyword list.
pattern matching against a list, assumes the order of the arguments.
with just one possible option it is fine, but not really flexibile.

skwerlman replied:
how can we make this a compile-time error then? i can see this becoming a problem later on

skwerlman replied:
i guess the other option is to derive a default value somehow from the `__MODULE__`

iggi42 replied:
yeah, I tried and failed

iggi42 replied:
I could be wrong, btw.
Maybe `__using__` does not enforce `args` datatype.
getting `use Client, __MODULE__` would mean calling a macro, to call a macro.

skwerlman replied:
i'm gonna merge this as-is. we can find a fix for this later. i'll open an issue for each pending change in this pr.

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,148 @@
+  alias Crawler
+
+  # macro to correct log debug messages
+  defmacrop debug(msg) do
```
maybe an issue for a different pr, but this probably deserves its own module.

actually, this would be a good candidate for the new `common` app we plan on adding

skwerlman commented on apps/crawler/lib/crawler/task_manager.ex:
```diff
@@ -0,0 +1,148 @@
+          manager :: GenServer.name(),
+          client :: module(),
+          timeout()
+        ) :: Client.search_result()
```
both `Client.query()` and `Client.search_result()` are undefined types

skwerlman reviewed (changes requested):
mostly minor changes needed, then we can get this in

Commit:
fix dialyzer warnings

skwerlman commented:
Once travis finishes (and there are no errors aside form the TODOs) this is good to merge :smiley:
